# Returning probabilities

Central to PennyLane is the QNode, or Quantum Node. This object combines:

A device <br>
A quantum function <br>
Any additional configuration information <br>

We first need a device. The device evaluates a quantum function. <br>

It could be either a simulator or actual quantum hardware. <br>

For the purpose of this coding challenge, we use "default.qubit", a simple built-in simulator that does not require external dependencies. <br>

To initialize a "default.qubit" device, we also need to specify the number of wires.

In [68]:
import json
import pennylane as qml
import pennylane.numpy as np

# Step 1: initialize a device
dev = qml.device("default.qubit", wires=2)

# Step 2: Add a decorator below - takes the function as input, spits out something new that we can evaluate
@qml.qnode(dev)
def simple_circuit(angle):
    """
    In this function:
        * Rotate the qubit around the x-axis by angle.
        * Measure the probability the qubit is in the zero state.

    Args:
        angle (float): how much to rotate a state around the x-axis.

    Returns:
        np.array(float): the probability of of the state being in the 0
        ground state.
    """

    # Step 3: Add gates to the QNode

    # Put your code here #
    qml.RX(angle, wires=0)

    # Step 4: Return the required probability 
    
    return qml.probs(wires=[0])


In [69]:
simple_circuit(1.45783)

tensor([0.55636311, 0.44363689], requires_grad=True)

In [71]:
simple_circuit(0.9572)

tensor([0.78790574, 0.21209426], requires_grad=True)

test_input: 1.45783
expected_output: 0.55636
test_input: 0.9572
expected_output: 0.78791

In [70]:
# These functions are responsible for testing the solution.
def run(test_case_input: str) -> str:
    angle = json.loads(test_case_input)
    output = simple_circuit(angle)[0]

    return str(output)


def check(solution_output: str, expected_output: str) -> None:
    solution_output = json.loads(solution_output)
    expected_output = json.loads(expected_output)
    assert np.allclose(solution_output, expected_output, rtol=1e-4)


# These are the public test cases
test_cases = [
    ('1.45783', '0.5563631060725739'),
    ('0.9572', '0.7879057356348377')
]

# This will run the public test cases locally
for i, (input_, expected_output) in enumerate(test_cases):
    print(f"Running test case {i} with input '{input_}'...")

    try:
        output = run(input_)

    except Exception as exc:
        print(f"Runtime Error. {exc}")

    else:
        if message := check(output, expected_output):
            print(f"Wrong Answer. Have: '{output}'. Want: '{expected_output}'.")

        else:
            print("Correct!")

Running test case 0 with input '1.45783'...
Correct!
Running test case 1 with input '0.9572'...
Correct!
